src/main/java/org/opensearch/flowframework/indices/FlowFrameworkIndicesHandler.java

```diff
@@ -732,41 +764,51 @@ private void getAndUpdateResourceInStateDocumentWithRetries(
                 updateRequest,
                 ActionListener.wrap(
-                    r -> handleStateUpdateSuccess(workflowId, newResource, listener),
-                    e -> handleStateUpdateException(workflowId, newResource, retries, listener, e)
+                    r -> handleStateUpdateSuccess(workflowId, resource, operation, listener),
+                    e -> handleStateUpdateException(workflowId, resource, operation, retries, listener, e)
                 )
             );
-        }, ex -> handleStateUpdateException(workflowId, newResource, 0, listener, ex)));
+        }, ex -> handleStateUpdateException(workflowId, resource, operation, 0, listener, ex)));
     }
 
-    private void handleStateUpdateSuccess(String workflowId, ResourceCreated newResource, ActionListener<WorkflowData> listener) {
+    private void handleStateUpdateSuccess(
+        String workflowId,
+        ResourceCreated newResource,
+        OpType operation,
+        ActionListener<WorkflowData> listener
+    ) {
         String resourceName = newResource.resourceType();
         String resourceId = newResource.resourceId();
         String nodeId = newResource.workflowStepId();
-        logger.info("Updated resources created for {} on step {} with {} {}", workflowId, nodeId, resourceName, resourceId);
+        logger.info(
+            "Updated resources created for {} on step {} to {} resource {} {}",
+            workflowId,
+            nodeId,
+            operation.equals(OpType.DELETE) ? "delete" : "add",
+            resourceName,
+            resourceId
+        );
         listener.onResponse(new WorkflowData(Map.of(resourceName, resourceId), workflowId, nodeId));
     }
 
     private void handleStateUpdateException(
         String workflowId,
         ResourceCreated newResource,
+        OpType operation,
         int retries,
         ActionListener<WorkflowData> listener,
         Exception e
     ) {
         if (e instanceof VersionConflictEngineException && retries > 0) {
             // Retry if we haven't exhausted retries
-            getAndUpdateResourceInStateDocumentWithRetries(workflowId, newResource, retries - 1, listener);
+            getAndUpdateResourceInStateDocumentWithRetries(workflowId, newResource, operation, retries - 1, listener);
             return;
         }
-        String errorMessage = "Failed to update workflow state for "
-            + workflowId
-            + " on step "
-            + newResource.workflowStepId()
-            + " with "
-            + newResource.resourceType()
-            + " "
-            + newResource.resourceId();
+        StringBuilder sb = new StringBuilder("Failed to update workflow state for ");
+        sb.append(workflowId).append(" on step ").append(newResource.workflowStepId());
+        sb.append(" to ").append(operation.equals(OpType.DELETE) ? "delete" : "add");
+        sb.append(" resource ").append(newResource.resourceType()).append(" ").append(newResource.resourceId());
```

Comment:
```suggestion
        String message = String.format(
    "Failed to update workflow state for %s on step %s to %s resource %s %s",
    workflowId,
    newResource.workflowStepId(),
    operation.equals(OpType.DELETE) ? "delete" : "add",
    newResource.resourceType(),
    newResource.resourceId()
);
```

Reply:
@owaiskazi19 `String.format()` has to parse the format string which [takes a couple dozen milliseconds](https://stackoverflow.com/a/44124045/1161484).  Granted this will be a rarely-encountered code path, but in addition to the added time, `String.format()` has to deal with locale (we have to specify locale or accept unpredictable output strings that can cause flaky tests), and can be a source of additional exceptions if you use the wrong format string, so I really like to avoid using it in exception catch blocks (which this `handle()` method is in).

While there's negligible difference between concatenation vs. `StringBuilder` I thought the way I separated the lines here was more readable than the concatenation version I replaced with it. 

The log message substitution stuff also works, but unfortunately since we're using the message both in an exception and log message, we have to do this stuff.

All that said, it looks like I can use the `ParameterizedMessageFactory` to construct this message in log4j2 style and then re-use it twice.   Which we should also do elsewhere in this repo...

Reply:
Implemented in https://github.com/opensearch-project/flow-framework/pull/898/commits/86ac8ea862ba59abcbdf036fd637159e751e3be4 and opened an issue to do elsewhere.